### PR TITLE
add hint (NO_)SWAP_JOIN_INPUTS

### PR DESCRIPTION
**Pull request by fzhedu**
### What problem does this PR solve? 
add hint (NO_)SWAP_JOIN_INPUTS,  SWAP_JOIN_INPUTS(table) means using the table to build hash table, and NO_SWAP_JOIN_INPUTS(table) means using the table to probe hash table.

### What is changed and how it works?
change the parser https://github.com/pingcap/parser/pull/652

### Check List 

Tests 

 - Unit test

Code changes

 - N/A

Side effects

 - N/A

Related changes

 - N/A

Release note

 - add hint (NO_)SWAP_JOIN_INPUTS like in Oracle

### CODE REVIEW

File: planner/core/exhaust_physical_plans.go [hunk 1]
```diff
@@ -1318,8 +1391,8 @@ func (p *LogicalJoin) exhaustPhysicalPlans(prop *property.PhysicalProperty) []Ph
 	}
 	joins = append(joins, indexJoins...)
 
-	hashJoins := p.getHashJoins(prop)
-	if (p.preferJoinType & preferHashJoin) > 0 {
+	hashJoins, forced := p.tryToGetHashJoins(prop)
+	if (p.preferJoinType&preferHashJoin) > 0 || forced {
```
Comment (SunRunAway) [inline planner/core/exhaust_physical_plans.go]: We may need treat `p.preferJoinType&preferHashJoin` as forced in `tryToGetHashJoins`.

File: executor/join_test.go [hunk 2]
```diff
@@ -127,6 +127,29 @@ func (s *testSuiteJoin2) TestJoin(c *C) {
 	result = tk.MustQuery("select * from t left outer join t1 on t.c1 = t1.c1 and t.c1 != 1 order by t1.c1")
 	result.Check(testkit.Rows("1 1 <nil> <nil>", "2 2 2 3"))
 
+	// Test SWAP_JOIN_INPUTS and NO_SWAP_JOIN_INPUTS for hash joins in outer joins
```
Comment (SunRunAway) [inline executor/join_test.go]: There're no `NO_SWAP_JOIN_INPUTS` tests here.

File: planner/core/exhaust_physical_plans.go [hunk 3]
```diff
@@ -257,26 +257,99 @@ func (p *PhysicalMergeJoin) initCompareFuncs() {
 		p.CompareFuncs = append(p.CompareFuncs, expression.GetCmpFunction(p.LeftJoinKeys[i], p.RightJoinKeys[i]))
 	}
 }
+func (p *LogicalJoin) tryToGetHashJoins(prop *property.PhysicalProperty) ([]PhysicalPlan, bool) {
```
Comment (SunRunAway) [inline planner/core/exhaust_physical_plans.go]: Could you add some related tests in `plan_suite_in.json`?

**Comment by XuHuaiyu** [comment]
Do the name and behavior of this hint be affirmed?

**Comment by SunRunAway** [comment]
We need this PR, reopen it.

**Comment by SunRunAway** [comment]
Something weird happens

```
mysql> desc select /*+ HASH_JOIN(t2), SWAP_JOIN_INPUTS(t2) */ * from testjoin t1 RIGHT JOIN testjoin t2 ON t1.a > t2.a;
+-------------------------+-------------+-----------+--------------------------------------------------------------------------------------------------------+
| id                      | count       | task      | operator info                                                                                          |
+-------------------------+-------------+-----------+--------------------------------------------------------------------------------------------------------+
| HashRightJoin_6         | 99900000.00 | root      | CARTESIAN right outer join, inner:TableReader_9, other cond:gt(randgen.testjoin.a, randgen.testjoin.a) |
| ├─TableReader_9         | 9990.00     | root      | data:Selection_8                                                                                       |
| │ └─Selection_8         | 9990.00     | cop[tikv] | not(isnull(randgen.testjoin.a))                                                                        |
| │   └─TableScan_7       | 10000.00    | cop[tikv] | table:t1, range:[-inf,+inf], keep order:false, stats:pseudo                                            |
| └─TableReader_11        | 10000.00    | root      | data:TableScan_10                                                                                      |
|   └─TableScan_10        | 10000.00    | cop[tikv] | table:t2, range:[-inf,+inf], keep order:false, stats:pseudo                                            |
+-------------------------+-------------+-----------+--------------------------------------------------------------------------------------------------------+
6 rows in set, 2 warnings (0.00 sec)

mysql> show warnings;
+---------+------+-------------------------------------------------------------------------------------------------------------------------------------------------+
| Level   | Code | Message                                                                                                                                         |
+---------+------+-------------------------------------------------------------------------------------------------------------------------------------------------+
| Warning | 1815 | There are no matching table names for (t2) in optimizer hint /*+ HASH_JOIN(t2) */ or /*+ TIDB_HJ(t2) */. Maybe you can use the table alias name |
| Warning | 1815 | Optimizer Hint /*+ SWAP_JOIN_INPUTS(t2) */ is inapplicable without column equal ON condition                                                    |
+---------+------+-------------------------------------------------------------------------------------------------------------------------------------------------+
2 rows in set (0.00 sec)
```

**Comment by fzhedu** [comment]
> HashRightJoin_6

this SQL actually takes CARTESIAN join, instead of hash joins, so the warning is right for current purpose which specifies input sides for hash joins. If we extend `SWAP_JOIN_INPUTS` for other joins, this warning is not correct , and we should let `SWAP_JOIN_INPUTS` work for CARTESIAN joins.

File: executor/join_test.go (hunk 2 above)
Comment (fzhedu) [reply executor/join_test.go]: ok

File: planner/core/exhaust_physical_plans.go [hunk 4]
```diff
@@ -257,26 +257,99 @@ func (p *PhysicalMergeJoin) initCompareFuncs() {
...
 	joins := make([]PhysicalPlan, 0, 2)
 	switch p.JoinType {
 	case SemiJoin, AntiSemiJoin, LeftOuterSemiJoin, AntiLeftOuterSemiJoin:
 		joins = append(joins, p.getHashJoin(prop, 1, false))
 	case LeftOuterJoin:
-		joins = append(joins, p.getHashJoin(prop, 1, false))
-		joins = append(joins, p.getHashJoin(prop, 1, true))
+		// TODO: support exchange the inner and the outer in CARTESIAN join
+		if len(p.EqualConditions) == 0 {
+			joins = append(joins, p.getHashJoin(prop, 1, false))
```
[the remark this answers is not in the document]
Comment (fzhedu) [reply planner/core/exhaust_physical_plans.go]: en, removed.

File: planner/core/exhaust_physical_plans.go [hunk 5]
```diff
@@ -257,26 +257,99 @@ func (p *PhysicalMergeJoin) initCompareFuncs() {
...
+			} else {
+				joins = append(joins, p.getHashJoin(prop, 0, false))
+				joins = append(joins, p.getHashJoin(prop, 0, true))
+			}
+		}
 	case InnerJoin:
-		joins = append(joins, p.getHashJoin(prop, 1, false))
-		joins = append(joins, p.getHashJoin(prop, 0, false))
+		if innerBuild {
+			joins = append(joins, p.getHashJoin(prop, 0, false))
```
[the remark this answers is not in the document]
Comment (fzhedu) [reply planner/core/exhaust_physical_plans.go]: No. Doing that will make the hint useless.

File: planner/core/logical_plan_builder.go [hunk 6]
```diff
@@ -391,7 +395,12 @@ func (p *LogicalJoin) setPreferredJoinType(hintInfo *tableHintInfo) {
 	if hintInfo.ifPreferMergeJoin(lhsAlias, rhsAlias) {
 		p.preferJoinType |= preferMergeJoin
 	}
-	if hintInfo.ifPreferHashJoin(lhsAlias, rhsAlias) {
+	// specify whether the inner be the build or probe side
+	if hintInfo.ifPreferHashJoinBuild(lhsAlias) || hintInfo.ifPreferHashJoinProbe(rhsAlias) {
+		p.preferJoinType |= preferHashJoinInnerBuild
```
[the remark this answers is not in the document]
Comment (fzhedu) [reply planner/core/logical_plan_builder.go]: ok

File: planner/core/exhaust_physical_plans.go (hunk 1 above)
Comment (fzhedu) [reply planner/core/exhaust_physical_plans.go]: move it into `tryToGetHashJoins`

**Comment by SunRunAway** [comment]
Closing it.

**Pull request closed without merging**